### DISCUSSION TITLE
How can I get the should-be-outputted text from a list of `Translation` objects?

### DISCUSSION THREAD

**Discussion by user202729** [opensteno/plover#1242] [Plugin Support] [answered]
... with `plover_build_utils.testing.CaptureOutput`?

It probably isn't that hard to get `_Action` objects, but parsing `*_attach` flags are not easy...

**Comment by user202729** [comment, marked as answer]
Okay this is the method...

Note that `Translation` objects does not handle macros (macros operate on `Translation` objects), use `Translation.translate_macro(macro)` or `Translation.translate_stroke(stroke)` instead if that's necessary.

Note: if you use Plover's `Translation` objects, make sure to not pass it directly `Formatter.format()` (instead, copy it with `Translation(translation.strokes, translation.english))`), so the `formatting` attribute is not destroyed/modified (which is important to ensure that the undo actions are correct).

Note: See also [How can I use Plover as a library from an independent script? · Discussion #1292 · openstenoproject/plover](https://github.com/openstenoproject/plover/discussions/1292) if you get some error.

```python
from plover_build_utils.testing import CaptureOutput
from plover.translation import Translation, Translator
from plover.formatting import Formatter
from plover.steno import Stroke

output = CaptureOutput()
formatter = Formatter(); formatter.set_output(output)
translator = Translator(); translator.add_listener(formatter.format)

# optional:
translator.set_min_undo_length(100)

# optional (set existing text, might affect the behavior of suffix):
output.text = ""

# optional (set dictionary to be used by the translator, if `translate_stroke` is used)
dictionary = translator.get_dictionary()
dictionary.set_dicts([StenoDictionary()])

# translate translations
translator.translate_translation(translation)

# alternatively (without undoing `translation.replaced`)
# NOTE private method, might change any time!
translator._do(translation)

# alternatively (no need for the translator)
# although then you will have to handle macros/context (prev) yourself
formatter.format(undo=[], do=[
	Translation(...),
	Translation(...),
	], prev=None)

# required if the translator is used
translator.flush()

# get output:

print(output.instructions)
# format: (NOTE might change between versions)
# ("b", int): backspace * number
# ("s", str): string
# ("c", str): key combination
# ("e", str): engine command

print(output.text)
```

**Reply by benoit-pierre** [reply]
```python
# alternatively (without undoing `translation.replaced`)
translator._do(translation)
```

No, private methods are off-limits.

If you want the same behavior as Plover, you need to do what the blackbox test framework in `plover_build_utils.testing` does.
